scripts/markdown/check-edit-links.js

```diff
@@ -0,0 +1,162 @@
+    batch.map(async ({ filePath, urlPath, editLink }) => {
+      try {
+        if (!editLink || ignoreFiles.some((ignorePath) => filePath.endsWith(ignorePath))) return null;
+
+        const controller = new AbortController();
+        const timeout = setTimeout(() => controller.abort(), TIMEOUT_MS);
+        const response = await fetch(editLink, {
+          method: 'HEAD',
+          signal: controller.signal
+        });
+        clearTimeout(timeout);
+        if (response.status === 404) {
+          return { filePath, urlPath, editLink };
+        }
+        return null;
+      } catch (error) {
+        return Promise.reject(new Error(`Error checking ${editLink}: ${error.message}`));
+      }
+    })
+  );
+}
+
+/**
+ * Check all URLs in batches
+ * @param {object[]} paths - Array of all path objects to check
+ * @returns {Promise<string[]>} Array of URLs that returned 404
+ */
+async function checkUrls(paths) {
+  const result = [];
+  const batchSize = Number(process.env.DOCS_LINK_CHECK_BATCH_SIZE) || 5;
+
+  const batches = [];
+  for (let i = 0; i < paths.length; i += batchSize) {
+    const batch = paths.slice(i, i + batchSize);
+    batches.push(batch);
+  }
+
+  console.log(`Processing ${batches.length} batches concurrently...`);
+  const batchResultsArray = await Promise.all(
+    batches.map(async (batch) => {
+      const batchResults = await processBatch(batch);
+      await pause(1000);
+      return batchResults.filter((url) => url !== null);
+    })
+  );
+
+  result.push(...batchResultsArray.flat());
+  return result;
+}
+
+/**
+ * Determines the appropriate edit link based on the URL path and file path
+ * @param {string} urlPath - The URL path to generate an edit link for
+ * @param {string} filePath - The actual file path
+ * @param {object[]} editOptions - Array of edit link options
+ * @returns {string|null} The generated edit link or null if no match
+ */
+function determineEditLink(urlPath, filePath, editOptions) {
+  // Remove leading 'docs/' if present for matching
+  const pathForMatching = urlPath.startsWith('docs/') ? urlPath.slice(5) : urlPath;
+
+  const target = editOptions.find((edit) => pathForMatching.includes(edit.value));
+
+  // Handle the empty value case (fallback)
+  if (target.value === '') {
+    return `${target.href}/docs/${urlPath}.md`;
+  }
+
+  // For other cases with specific targets
+  return `${target.href}/${path.basename(filePath)}`;
+}
+
+/**
+ * Recursively processes markdown files in a directory to generate paths and edit links
+ * @param {string} folderPath - The path to the folder to process
+ * @param {object[]} editOptions - Array of edit link options
+ * @param {string} [relativePath=''] - The relative path for URL generation
+ * @param {object[]} [result=[]] - Accumulator for results
+ * @returns {Promise<object[]>} Array of objects containing file paths and edit links
+ */
+async function generatePaths(folderPath, editOptions, relativePath = '', result = []) {
```

Comment:
Consider optimizing directory traversal.

The current implementation reads the entire directory tree synchronously. For large documentation repositories, this could be optimized.

Consider using a streaming directory reader like readdir-enhanced or implementing a generator-based solution:

```
async function* walkDirectory(dir) {
  const files = await fs.readdir(dir);
  for (const file of files) {
    const filePath = path.join(dir, file);
    const stats = await fs.stat(filePath);
    if (stats.isDirectory()) {
      yield* walkDirectory(filePath);
    } else {
      yield filePath;
    }
  }
}
```

Reply:
@coderabbitai Let's create an issue for this and we will target this in next PR.